services/scheduler/scheduler.go

```diff
@@ -0,0 +1,218 @@
+	s.scheduler.Stop()
+}
+
+// Add adds job to scheduler and save it to DB.
+func (s *Service) Add(job Job, cronExpr string, startAt time.Time, retry uint, retryInterval time.Duration) (*models.ScheduleJob, error) {
+	var dbJob *models.ScheduleJob
+	var err error
+
+	err = s.db.InTransaction(func(tx *reform.TX) error {
+		j := s.scheduler.Cron(cronExpr).SingletonMode()
+		if !startAt.IsZero() {
+			j = j.StartAt(startAt)
+		}
+
+		dbJob, err = models.CreateScheduleJob(tx.Querier, models.CreateScheduleJobParams{
+			CronExpression: cronExpr,
+			StartAt:        startAt,
+			Type:           job.Type(),
+			Data:           job.Data(),
+			Retries:        retry,
+			RetryInterval:  retryInterval,
+		})
+		if err != nil {
+			return err
+		}
+
+		id := dbJob.ID
+		fn := s.wrapJob(job, id, int(retry), retryInterval)
+
+		scheduleJob, err := j.Tag(id).Do(fn)
+		if err != nil {
+			return err
+		}
+
+		dbJob, err = models.ChangeScheduleJob(tx.Querier, id, models.ChangeScheduleJobParams{
+			NextRun: scheduleJob.NextRun(),
+			LastRun: scheduleJob.LastRun(),
+		})
+		if err != nil {
+			s.l.WithField("id", id).Errorf("failed to set next run for new created job")
+		}
+
+		return nil
+	})
+	return dbJob, err
+}
+
+// Remove stops job specified by id and removes it from DB and scheduler.
+func (s *Service) Remove(id string) error {
+	s.jobsMx.RLock()
+	if cancel, ok := s.jobs[id]; ok {
+		cancel()
+	}
+	s.jobsMx.RUnlock()
+	err := s.scheduler.RemoveByTag(id)
+	if err != nil {
+		return err
+	}
+
+	if err := models.RemoveScheduleJob(s.db.Querier, id); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) loadFromDB() error {
+	s.scheduler.Clear()
+	// Reset tags
+	s.scheduler.TagsUnique()
+
+	disabled := false
+	jobs, err := models.FindScheduleJobs(s.db.Querier, models.ScheduleJobsFilter{
+		Disabled: &disabled,
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, dbJob := range jobs {
+		job, err := s.convertDBJob(dbJob)
+		if err != nil {
+			return err
+		}
+		fn := s.wrapJob(job, dbJob.ID, int(dbJob.Retries), dbJob.RetryInterval)
+		j := s.scheduler.Cron(dbJob.CronExpression).SingletonMode()
+		if !dbJob.StartAt.IsZero() {
+			j = j.StartAt(dbJob.StartAt)
+		}
+		if _, err := j.Tag(dbJob.ID).Do(fn); err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+func (s *Service) wrapJob(job Job, id string, retry int, retryInterval time.Duration) func() {
+	return func() {
+		var err error
+		l := s.l.WithField("jobType", job.Type())
+		ctx, cancel := context.WithCancel(context.Background())
+
+		s.jobsMx.Lock()
+		s.jobs[id] = cancel
+		s.jobsMx.Unlock()
+
+		defer func() {
+			cancel()
+			s.jobsMx.Lock()
+			delete(s.jobs, id)
+			s.jobsMx.Unlock()
+		}()
+
+		for {
+			t := time.Now()
+			l.Debug("Starting job")
+			err = job.Do(ctx)
+			l.WithField("duration", time.Since(t)).Debug("Ended  job")
+			if err == nil || err == context.Canceled {
+				break
+			} else {
+				l.Error(err)
+			}
+			if retry <= 0 {
+				break
+			}
+			retry--
```

Comment:
I wonder, do we need to track retries id DB? Because now we will lose retries counter in case of a server restart.

Reply:
Good question. Now I hold retries just for new schedule job. So, add also current retry attemp?

Reply:
yes, I think we need current retry attempt in the DB.

Reply:
But does it make sense to continue retries when the server was shut down for 2 days, for example?